### PR TITLE
fix:Fix request data in async task issue

### PR DESCRIPTION
**Pull request by tomaszpowroznik**
Currently RequestData accessed from @Async executor may be corrupted (returns nulls) when main request thread already committed http response (depends on timing). It can affect payments, share a case and notifications.
New approach is doing copy of RequestData in main request thread and attach it to every new thread spanned as ThreadLocal variable so accessing it from executor works fine even if http requested is closed.

### CODE REVIEW

File: service/src/main/java/uk/gov/hmcts/reform/fpl/request/CacheAwareRequestData.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+@Primary
+@Service
+public class CacheAwareRequestData implements RequestData {
+
+    private final HttpServletRequest httpServletRequest;
+
+    @Autowired
+    public CacheAwareRequestData(HttpServletRequest httpServletRequest) {
+        this.httpServletRequest = httpServletRequest;
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/request/CacheAwareRequestData.java:17]: Does this servlet request still have a possibility of getting cleaned up before we have used it?

Comment (tomaszpowroznik) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/request/CacheAwareRequestData.java:17]: If RequestData is accessed from async task then AsyncConfiguration guarantees that there will be cached version of RequestData and httpRequest wont be consulted.
If Request data is accessed from standard "request"  thread then it is guaranteed by spring request scope that httpServletRequest is valid.
The question is about when exactly cached version is created, and it created in request thread.

 @Override
        public Runnable decorate(@Nonnull Runnable task) {
            //request thread where we create copy of request data
            SimpleRequestData requestData = new SimpleRequestData(context.getBean(RequestData.class));

           //everything in lambda is executed in new task thread
            return () -> {
                RequestDataCache.add(requestData);
                try {
                    task.run();
                } finally {
                    RequestDataCache.remove();
                }
            };
        }

Comment (tomaszpowroznik) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/request/CacheAwareRequestData.java:17]: To test it i have setup code like this:
    @PostMapping("/submitted")
    public void handleSubmittedEvent(@RequestBody CallbackRequest callbackRequest) {
        System.out.println("Invoking async action " + Thread.currentThread().getName());
        localAuthorityUserService.grantUserAccessWithCaseRole(caseId, caseLocalAuthority);
    }

in AsyncConfiguration
 @Override
        public Runnable decorate(@Nonnull Runnable task) {
            System.out.println("Copying request data " + Thread.currentThread().getName());
            SimpleRequestData requestData = new SimpleRequestData(context.getBean(RequestData.class));

            return () -> {
                System.out.println("Setting thread local"+Thread.currentThread().getName());
                RequestDataCache.add(requestData);
                try {
                    task.run();
                } finally {
                    RequestDataCache.remove();
                }
            };
        }

output is: 
2020-06-24T17:32:15.734 INFO  [http-nio-4000-exec-1] u.g.h.r.logging.filters.RequestStatusLoggingFilter Request POST /callback/case-initiation/about-to-submit processed in 242ms
Invoking async action http-nio-4000-exec-2
Copying request data http-nio-4000-exec-2
Setting thread localgetAsyncExecutor-1

which proves that copy of request data is done in request thread where correct httpRequest is guranteed by spring

Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/request/CacheAwareRequestData.java:25]: Sounds good to me.
On a slight side note if you care, you can change the name from `localgetAsyncExecutor` to something else in the `@Bean(name="something")`

File: service/src/test/java/uk/gov/hmcts/reform/fpl/request/RequestDataTest.java [hunk 2]
```diff
@@ -9,28 +12,63 @@
...
     @InjectMocks
-    private RequestData requestData;
+    private CacheAwareRequestData requestData;
 
-    @Test
-    public void shouldReturnAuthorisationHeaderFromRequest() {
-        when(httpServletRequest.getHeader("authorization")).thenReturn(AUTH_TOKEN);
-        assertThat(requestData.authorisation()).isEqualTo(AUTH_TOKEN);
+    @Nested
+    class NotCachedRequestData {
```
Comment (lewisbirks) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/request/RequestDataTest.java:32]: test methods don't need to be public

File: service/src/main/java/uk/gov/hmcts/reform/fpl/config/AsyncConfiguration.java [hunk 3]
```diff
@@ -35,20 +41,28 @@ public void handleUncaughtException(Throwable throwable, Method method, Object..
     @Bean
     public Executor getAsyncExecutor() {
         ThreadPoolTaskExecutor taskExecutor = new ThreadPoolTaskExecutor();
-        taskExecutor.setTaskDecorator(new AsyncTaskDecorator());
+        taskExecutor.setTaskDecorator(new AsyncTaskDecorator(context));
```
[the remark this answers is not in the document]
Comment (tomaszpowroznik) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/config/AsyncConfiguration.java:44]: cannot be injected into method, this method signature is enforced by interface

File: service/src/test/java/uk/gov/hmcts/reform/fpl/request/RequestDataTest.java (hunk 2 above)
Comment (tomaszpowroznik) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/request/RequestDataTest.java:32]: fixed

File: service/src/main/java/uk/gov/hmcts/reform/fpl/config/AsyncConfiguration.java [hunk 4]
```diff
@@ -18,6 +21,9 @@
 @Configuration
 public class AsyncConfiguration implements AsyncConfigurer {
 
+    @Autowired
+    private ApplicationContext context;
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/config/AsyncConfiguration.java:25]: Do we want to have field injection or is constructor injection better?

Comment (tomaszpowroznik) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/config/AsyncConfiguration.java]: yeah, for consistency it will be better to change it

**Pull request merged**